File: LC50/lc15.py
"""
Given an array S of n integers, are there elements a, b, c in S
 such that a + b + c = 0?
Find all unique triplets in the array which gives the sum of zero.

Note: The solution set must not contain duplicate triplets.

For example, given array S = [-1, 0, 1, 2, -1, -4],

A solution set is:
[
  [-1, 0, 1],
  [-1, -1, 2]
]
"""


def three_sum(nums):
    output = []
    nums.sort()
    for i in range(len(nums) - 2):
        if i != 0 and nums[i] == nums[i - 1]:
            continue

        first = nums[i]
        target = -first

        j = i + 1
        k = len(nums) - 1
        while j < k:
            second = nums[j]
            third = nums[k]

            if second + third == target:
                output.append([first, second, third])
                j += 1
                while j < k and nums[j + 1] == second:
                    j += 1

                k -= 1
                while j < k and nums[k - 1] == third:
                    k -= 1
            elif second + third < target:
                j += 1
            else:
                k -= 1

    return output


def test():
    nums_list = [
        [0, 0, 0],
        [-1, 0, 1, 2, -1, -4]
    ]
    for nums in nums_list:
        print(nums)
        print(three_sum(nums))
        print("\n")


test()

File: LC50/lc15.js
function three_sum(nums) {
    nums.sort();
    var output = [],
        first, second, third;
    for (var i = 0; i < nums.length - 2; i++) {
        if (i !== 0 && nums[i] === nums[i-1]) {
            continue;
        }
        first = nums[i];
        target = -first;
        var j = i + 1;
        var k = nums.length - 1;
        while (j < k) {
            second = nums[j];
            third = nums[k];
            if (second + third === target) {
                output.push([first, second, third]);
                j += 1;
                k -= 1;
                while (j < k && nums[j] == second) {
                    j += 1;
                }
                while (j < k && nums[k] == third) {
                    k -= 1;
                }
            } else if (second + third < target) {
                j += 1;
            } else {
                k -= 1;
            }
        }
    }
    return output;
}


function test() {
    var nums_list = [
        [0, 0, 0],
        [-1, 0, 1, 2, -1, -4]
    ];
    for (var nums of nums_list) {
        console.log(nums);
        console.log(three_sum(nums));
    }
}


test();
